File: src/main/go/ch6/look-and-say.go
package main

func main() {
	c := ant(10)
	for value := range c {
		print(value)
	}
	println()
}

func ant(n int) <-chan int {
	c := gen(1)
	for i := 0; i < n; i++ {
		c = next(c)
	}
	return c
}

func gen(n int) <-chan int {
	c := make(chan int)
	go func() {
		c <- n
		close(c)
	}()
	return c
}

func next(in <-chan int) <-chan int {
	out := make(chan int)
	go func() {
		prev, count := <-in, 1
		for value := range in {
			if value == prev {
				count++
			} else {
				out <- count
				out <- prev
				prev, count = value, 1
			}
		}
		out <- count
		out <- prev
		close(out)
	}()
	return out
}

File: src/main/js/ch6/look-and-say.js
'use strict'

const READ = {type: 'read'}
const WRITE = (value) => ({type: 'write', value})

function* next() {
  let prev = yield READ
  let count = 1
  let value
  while (value = yield READ) {
    if (prev === value) {
      count++
    } else {
      yield WRITE(count)
      yield WRITE(prev)
      prev = value
      count = 1
    }
  }
  yield WRITE(count)
  yield WRITE(prev)
}

function* ant(n) {
  let procs = new Array(n + 1)
  procs[0] = function* () { yield WRITE(1) } ()
  for (let i = 1; i < n+1; i++) {
    procs[i] = next()
  }
  yield* dispatch(procs)
}

function* dispatch(procs) {
  let value
  let cur = n 
  while (true) {
    let next = procs[cur].next(value)
    if (next.done) {
      if (cur === n) {
        return
      } else {
        value = undefined
        cur++
      }
    } else {
      if (next.value.type === 'read') {
        cur--
      } else if (cur === n) {
        yield next.value.value
      } else {
        value = next.value.value
        cur++
      }
    }
  }
}

const n = process.argv.length >= 3 ? parseInt(process.argv[2]) : 10
const m = process.argv.length >= 4 ? parseInt(process.argv[3]) : 100
let count = 0

for (let v of ant(n)) {
  if (count++ >= m) break
  process.stdout.write(`${v}`)
}
console.log()